beacon-chain/powchain/options.go

```diff
@@ -0,0 +1,88 @@
+package powchain
+
+import (
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/prysmaticlabs/prysm/beacon-chain/cache/depositcache"
+	statefeed "github.com/prysmaticlabs/prysm/beacon-chain/core/feed/state"
+	"github.com/prysmaticlabs/prysm/beacon-chain/db"
+	"github.com/prysmaticlabs/prysm/beacon-chain/state/stategen"
+	"github.com/prysmaticlabs/prysm/network"
+)
+
+type Option func(s *Service) error
+
+// WithHttpEndpoints deduplicates and parses http endpoints for the powchain service to use,
+// and sets the "current" endpoint that will be used first.
+func WithHttpEndpoints(endpointStrings []string) Option {
+	return func(s *Service) error {
+		stringEndpoints := dedupEndpoints(endpointStrings)
+		endpoints := make([]network.Endpoint, len(stringEndpoints))
+		for i, e := range stringEndpoints {
+			endpoints[i] = HttpEndpoint(e)
+		}
+		// Select first http endpoint in the provided list.
+		var currEndpoint network.Endpoint
+		if len(endpointStrings) > 0 {
+			currEndpoint = endpoints[0]
+		}
+		s.cfg.httpEndpoints = endpoints
+		s.cfg.currHttpEndpoint = currEndpoint
+		return nil
+	}
+}
+
+// WithDepositContractAddress for the deposit contract.
+func WithDepositContractAddress(addr common.Address) Option {
+	return func(s *Service) error {
+		s.cfg.depositContractAddr = addr
+		return nil
+	}
+}
+
+// WithDatabase for the beacon chain database.
+func WithDatabase(database db.HeadAccessDatabase) Option {
```

Comment:
Unfortunately a little duplication is better than the wrong abstraction here. It has to return powchain.Option because it is mutating the actual powchain struct. We can't make this generic without some weird abstractions. Also, it makes it clear from godoc that the powchain service _has_ a `WithDatabase` option once can use. It adds to documentation

Reply:
I forgot about the different return type. You're 100% correct.